addons/mass_mailing/models/mail_notification.py

```diff
@@ -49,53 +40,59 @@ class MailingTrace(models.Model):
     links_click_ids = fields.One2many('link.tracker.click', 'mailing_trace_id', string='Links click')
     clicked = fields.Datetime(help='Date when customer clicked on at least one tracked link')
     # Status
-    state = fields.Selection(compute="_compute_state",
-                             selection=[('outgoing', 'Outgoing'),
-                                        ('sent', 'Sent'),
-                                        ('opened', 'Opened'),
-                                        ('replied', 'Replied'),
-                                        ('bounced', 'Bounced'),
-                                        ('exception', 'Exception'),
-                                        ('canceled', 'Canceled')], store=True)
-    failure_type = fields.Selection(selection=[
-        ("SMTP", "Connection failed (outgoing mail server problem)"),
-        ("RECIPIENT", "Invalid email address"),
-        ("BOUNCE", "Email address rejected by destination"),
-        ("UNKNOWN", "Unknown error")], string='Failure type')
-    state_update = fields.Datetime(compute="_compute_state", string='State Update',
-                                   help='Last state update of the mail',
-                                   store=True)
-
-    @api.depends('trace_type', 'mass_mailing_id')
+    notification_status = fields.Selection(selection_add=[
+        ('outgoing', 'Outgoing'),
+        ('opened', 'Opened'),
+        ('replied', 'Replied')])
+    status_update = fields.Datetime(string='State Update', help='Last state update of the mail', store=True)
+
+    @api.depends('mass_mailing_id.campaign_id')
+    def _compute_campaign_id(self):
+        for notif in self:
+            notif.campaign_id = notif.mass_mailing_id.campaign_id
+
+
+    @api.depends('notification_type', 'mass_mailing_id', 'res_partner_id')
     def _compute_display_name(self):
```

Comment:
While you're at it, remove the `display_name` field override and implement a proper `name_get` instead.